File: app/src/main/java/com/boardgamegeek/sorter/MoneySorter.java
package com.boardgamegeek.sorter;

import android.content.Context;
import android.database.Cursor;
import android.support.annotation.NonNull;
import android.text.TextUtils;

import com.boardgamegeek.util.CursorUtils;
import com.boardgamegeek.util.PresentationUtils;

public abstract class MoneySorter extends CollectionSorter {
	public static final String MISSING_DATA = "-";

	public MoneySorter(Context context) {
		super(context);
		orderByClause = getCurrencyColumnName() + " DESC, " + getAmountColumnName() + " DESC";
	}

	@NonNull
	protected abstract String getAmountColumnName();

	@NonNull
	protected abstract String getCurrencyColumnName();

	@Override
	public String[] getColumns() {
		return new String[] { getCurrencyColumnName(), getAmountColumnName() };
	}

	@NonNull
	@Override
	public String getDisplayInfo(@NonNull Cursor cursor) {
		double amount = CursorUtils.getDouble(cursor, getAmountColumnName());
		String currency = getString(cursor, getCurrencyColumnName());
		String info = PresentationUtils.describeMoney(currency, amount);
		return getInfoOrMissingInfo(info);
	}

	@Override
	public String getHeaderText(@NonNull Cursor cursor) {
		double amount = round(CursorUtils.getDouble(cursor, getAmountColumnName()));
		String currency = getString(cursor, getCurrencyColumnName());
		String info = PresentationUtils.describeMoneyWithoutDecimals(currency, amount);
		return getInfoOrMissingInfo(info);
	}

	@NonNull
	private String getInfoOrMissingInfo(String info) {
		if (TextUtils.isEmpty(info)) {
			return MISSING_DATA;
		}
		return info;
	}

	private double round(double value) {
		return ((int) (Math.ceil(value + 9) / 10)) * 10;
	}
}

File: app/src/main/java/com/boardgamegeek/sorter/MoneySorter.kt
package com.boardgamegeek.sorter

import android.content.Context
import com.boardgamegeek.entities.CollectionItem
import com.boardgamegeek.extensions.asMoney
import java.text.DecimalFormat
import kotlin.math.ceil

abstract class MoneySorter(context: Context) : CollectionSorter(context) {
    protected abstract fun amount(item: CollectionItem): Double
    protected abstract fun currency(item: CollectionItem): String

    override fun sortAscending(items: Iterable<CollectionItem>): List<CollectionItem> {
        return items.sortedWith(compareBy<CollectionItem> { currency(it) }.thenBy { amount(it) })
    }

    override fun sortDescending(items: Iterable<CollectionItem>): List<CollectionItem> {
        return items.sortedWith(compareByDescending<CollectionItem> { currency(it) }.thenByDescending { amount(it) })
    }

    private fun round(value: Double) = (ceil(value / 10).toInt() * 10).toDouble()

    override fun getHeaderText(item: CollectionItem) = round(amount(item)).asMoney(currency(item), DecimalFormat("0")).ifEmpty { MISSING_DATA }

    override fun getDisplayInfo(item: CollectionItem) = amount(item).asMoney(currency(item)).ifEmpty { MISSING_DATA }

    companion object {
        private const val MISSING_DATA = "-"
    }
}
